### DISCUSSION TITLE
SAAS Permissions

### DISCUSSION THREAD

**Discussion by Attacler** [directus/directus#7054] [Feature Requests]
Hi there,

So for a SAAS product i need permissions for each user in each organisation.
What i think is good to do is to add a extra layer to the permissions.
Basicly so every user has differend permissions when they try do to something within another organisation.

this topic is up for debate since it has to be functional and also programmable.

Current idea (from comments):
     Allow multiple roles for 1 user
     1 role has to be the default
     when calls are made, you can provide another role in order to make use that

My extra idea:
     When a user has more than 1 role, show a dropdown somewhere in the Directus UI to select another role
     Roles should have a "display value", so that you can give it a differend name.
     For example instead of displaying "Orgname - rolename" you can make it just appear as "Orgname"

**Comment by rijkvanzanten** [comment]
> Basicly so every user has differend permissions when they try do to something within another organisation.

Could you provide a more specific real-life example? 

Would this be solved with the ability to have multiple roles per user? (something that has come up time and again in the past)

**Reply by Attacler** [reply]
> > Basicly so every user has differend permissions when they try do to something within another organisation.
> 
> Could you provide a more specific real-life example?
> 
> Would this be solved with the ability to have multiple roles per user? (something that has come up time and again in the past)

So in my senario we have the following:
organisations
users

for example a user can be "developer" in one organisation, thus have less permissions than a administrator
But that same user can be a administrator elsewhere.

so we have to somehow have to make sure people can for example view logs with role one in organisation one but cannot read those in organisation 2. but is able to read and edit those in organisation 3.

because its saas, i dont want to create roles for each organisation.

so in my case how it would work:
user requests something.
organisation role is getting picked.
request is completed with that role.

**Reply by rijkvanzanten** [reply]
> because its saas, i dont want to create roles for each organisation.

Could you elaborate on this? Regardless of how we end up structuring this, you'll have to have some sort of way to group these users within their respective organizations right? Otherwise there's no way to know whether or not they'd have access to something 🤔 

> Would this be solved with the ability to have multiple roles per user?

The reason I mentioned this is because you could have a structure where you have roles for:

```
1. Org 1 — Admin
2. Org 1 — Developer

3. Org 2 — Admin
4. Org 2 — Developer
```

where a user would be part of roles 1 and 4, giving them admin access to everything in org 1, while only getting dev access for the stuff in org 2

**Reply by Attacler** [reply]
> > because its saas, i dont want to create roles for each organisation.
> 
> Could you elaborate on this? Regardless of how we end up structuring this, you'll have to have some sort of way to group these users within their respective organizations right? Otherwise there's no way to know whether or not they'd have access to something 🤔
> 
> > Would this be solved with the ability to have multiple roles per user?
> 
> The reason I mentioned this is because you could have a structure where you have roles for:
> 
> ```
> 1. Org 1 — Admin
> 2. Org 1 — Developer
> 
> 3. Org 2 — Admin
> 4. Org 2 — Developer
> ```
> 
> where a user would be part of roles 1 and 4, giving them admin access to everything in org 1, while only getting dev access for the stuff in org 2

Yes but right now you only can select 1 role at a time.
which means that whenever someone has 2 tabs open (org 1 and org2) it would messup.

**Reply by rijkvanzanten** [reply]
> Yes but right now you only can select 1 role at a time.

Yeah! I was asking whether or not having support for multiple roles per user would solve for this particular use case, which it seems it would 🙂

**Reply by Attacler** [reply]
> > Yes but right now you only can select 1 role at a time.
> 
> Yeah! I was asking whether or not having support for multiple roles per user would solve for this particular use case, which it seems it would 🙂

yes!
so now the question is, how do we solve multiple roles 😂

**Reply by rijkvanzanten** [reply]
haha yeah, that's the real big question here innit

**Reply by Attacler** [reply]
> haha yeah, that's the real big question here innit

There are multiple approaches:
1. You basicly execute a call based on all the roles of a user. kinda expensive and not necessary.
2. You give a user a "default" role, whenever you make a call, you can give the desired role id with it.
    if its empty, the default role will be used, if its filled, it has to be checked if the role is indeed there.
3. Make organisations a default but you would still need number 1 or 2

**Reply by rijkvanzanten** [reply]
Or option 4, use an `_or` to merge both sets of permissions, as that effectively means "one with the most permission is used"

**Comment by bernatvadell** [comment]
It's very interesting, we could probably do something similar to this and I don't think it would make a "dramatic" change:

1) Add an option to enable "Multi-organization" projects in the envs? ex: PROJECT_TYPE="saas"

Doing this would create the following entities / collections:
- directus_organizations (id, name, is_system, ...)
- directus_user_organizations (user_id, organization_id, role_id)

I would modify the "directus_users" collection to add "organization_id_active" or similar

2) When the user identifies himself:
a) It is checked if the user has an active organization and if so, it is included in the JWT.
b) If the user does not have an active organization, but has some organization assigned (directus_user_organizations), the first one is assigned and it is included in the JWT. (This is like the last selected organization. Although in a first access, you will not have any.)

3) By enabling this option, where the name of the project currently appears, I would change it to be the name of the organization, and in the event that a user had more than one, it would show him a dropdown to be able to change.

Changing would update the value of directus_users.organization_id_active and generate a new JWT. (In the next login, this will be the option selected by default.)

![image](https://user-images.githubusercontent.com/17298604/127387018-59456c91-a376-4d30-af34-96cc6ce46221.png)


4) In the same way that we currently control through permissions with: $CURRENT_USER, $CURRENT_ROLE, I would add: $CURRENT_ORGANIZATION (This value will be indicated by the JWT. )


In this way, the Directus core would work the same, that is, the possibility that a user can view all the records to which they have access in their multiple organizations is not contemplated.

If this were the case, it would complicate the core, suppose we have different roles in several organizations, in one role you have visibility to some columns and in another role to others. If we show all the records of the different organizations, in the filters, selection of visible columns, etc ... What columns do we show?

In short, the entire core would continue to work the same, the only thing we would do would be to manage an update of the JWT to include the selected organization. 


Sorry for my english.

**Comment by aidenfoxx** [comment]
For me, allowing users to have multiple roles feels like the much more robust and elegant solution. Some can benefit from splitting user into organizations, but most people can benefit from multi-role support.

**Reply by rijkvanzanten** [reply]
I would agree, less proprietary/arbitrary that way

**Reply by bernatvadell** [reply]
If we define the path, I would love to get the job done.

**Reply by rijkvanzanten** [reply]
Lets keep discussing on https://github.com/directus/directus/discussions/4959. That being said, this is a _big deal_ so I'd rather not just jump ahead without having a very clear understanding of the goals and processes involved

**Reply by bernatvadell** [reply]
true, but in this case, I think we are not talking about a user can have multiple roles, but rather, link a role of the user to an organization. 

![image](https://user-images.githubusercontent.com/17298604/127391724-87dab4cd-c3a3-4fac-99c4-a105871b704a.png)

**Reply by aidenfoxx** [reply]
> true, but in this case, I think we are not talking about a user can have multiple roles, but rather, link a role of the user to an organization.

That goal can be achieved through a user having multiple roles as rijk mentioned earlier:

> ```
> 1. Org 1 — Admin
> 2. Org 1 — Developer
> 
> 3. Org 2 — Admin
> 4. Org 2 — Developer
> ```
> where a user would be part of roles 1 and 4, giving them admin access to everything in org 1, while only getting dev access for > the stuff in org 2

 Anything past that doesn't follow the 80/20 rule.
 
 > this is a big deal so I'd rather not just jump ahead without having a very clear understanding of the goals and processes involved
 
 I can't tell if this is a dig at my authentication changes 😂

**Reply by rijkvanzanten** [reply]
> I can't tell if this is a dig at my authentication changes 😂

haha no not at all!

Having multiple roles per user opens up a bigger can of worms than we're thinking about right now. One thing that comes to mind immediately is that we'd need to have a way to show the full permissions available to a single user based on their roles somewhere in the app, otherwise it'll be impossible to keep track who has access to what as soon as a single user is in more than 2 roles!

**Reply by benhaynes** [reply]
And would need to combine things like Nav/Module overrides. 😬

**Reply by rijkvanzanten** [reply]
Yeah exactly, and anything else role specific, like IP address allow list

**Reply by bernatvadell** [reply]
I think adding support for multiple roles (simultaneously) would be a huge job.  In the example of the project that I am developing, there is no possibility that a user has several roles, which is possible, is that he belongs to several organizations and has a specific role for each of them.  By having an organization selector, the change is not that great since your entire application works with the selected organization, including the role.

**Reply by bernatvadell** [reply]
I think they are different things that can be addressed independently.

**Reply by rijkvanzanten** [reply]
Sure, but multi-tenancy in the form of standalone organizations with scoped app instances is not something I think we're looking to implement at this time. We used to have a very similar setup in v7/v8, and the amount of maintenance overhead outweighed the usability for a tiny part of the user base by _a lot_. Support for multiple roles is something that has come up time and time again, and it would also provide the functionality required for this particular use case 👍🏻

**Reply by bernatvadell** [reply]
I would like to emphasize that I have not talked about multiple instances.  only from a selector that would add a claim to the jwt.  it is true that with multiple roles x user, it could cover this need.  however, I think it would be a headache to manage roles as if they were organizations.  taking into account that the registration of the organization would be done by the user himself from a frontend, I understand that through a hook the roles could be registered, but if I want to make a change in any of the roles, I should make a bulk change in the existing roles and modify the hook for the new ones.

That said, I agree that this only solves the problem for a small group of applications.  probably for this specific project, it should be a bespoke solution.

**Comment by tspvivek** [comment]
I had the same requirement for SAAS and i modified authenticate code logic and added $CURRENT_ORGANISATION and $CURRENT_DEPARTMENT variables  and used them in permissions. But every time i update directus i need to add my custom code changes back in authenticate.js.

I am sure this can be added but need some UI for users to switch between his roles.

**Reply by aidenfoxx** [reply]
You'll probably like the new changes to authentication then. It's been updated to allow multiple different login options, each of which store a `provider` value in the users table which links a user to a specific login flow. So theoretically you could create different logins for each organization and use the `provider` value for your customization.

**Reply by bernatvadell** [reply]
I suppose it would be possible, but in the ideal scenario, just registering the organization and linking a user, we should already have the system working.  anything else, will imply an overhead every time we register a new organization.

**Comment by imiskolee** [comment]
linked another same thing: https://github.com/directus/directus/discussions/14115

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
